### PR TITLE
cross platform: enable Runtime/Types

### PR DESCRIPTION
**Pull request by obastemur**
This PR enables most of the files under `Runtime/Types` for cross platform compilation. Gonna work on the rest while this chunk is being reviewed and merged. 

Most of the changes are due to issues with Clang's support of C++11 template etc. features.

### CODE REVIEW

File: lib/Runtime/Types/TypePropertyCache.cpp [hunk 1]
```diff
@@ -118,7 +118,7 @@ namespace Js
         Assert(prototypeObjectWithProperty);
 
         const TypePropertyCacheElement &element = elements[ElementIndex(id)];
-        if(element.Id() != id || !checkMissing && element.IsMissing())
+        if(element.Id() != id || (!checkMissing && element.IsMissing()))
```
Comment (obastemur) [inline lib/Runtime/Types/TypePropertyCache.cpp]: Edited this one in order to get rid of a clang warning. Hoping the location for parentheses are correct?

File: lib/Runtime/Types/RecyclableObject.cpp [hunk 2]
```diff
@@ -448,7 +448,9 @@ namespace Js
         {
             //In ES5 mode strict equals (===) on same instance of object type VariantDate succeeds.
             //Hence equals needs to succeed.
-            goto ReturnTrue;
+            //goto ReturnTrue;
+            *value = TRUE;
+            return TRUE;
```
Comment (obastemur) [inline lib/Runtime/Types/RecyclableObject.cpp]: Clang is unhappy with this goto skipping the initialization of scope variables. I can also remove the comment?

File: lib/Runtime/Types/CMakeLists.txt [hunk 3]
```diff
@@ -11,17 +8,17 @@ add_library (Chakra.Runtime.Types
     DynamicObjectSnapshotEnumerator.cpp
     DynamicObjectSnapshotEnumeratorWPCache.cpp
```
Comment (digitalinfinity) [inline lib/Runtime/Types/CMakeLists.txt]: This file still has build errors for me- please comment it out and fix it in the next commit

Comment (obastemur) [inline lib/Runtime/Types/CMakeLists.txt]: @digitalinfinity I had tested with Ubuntu 15/Clang 3.6 . Can you please share some details?
```
./build.sh CXX=/usr/bin/clang++-3.6 CC=/usr/bin/clang-3.6
```

Comment (digitalinfinity) [inline lib/Runtime/Types/CMakeLists.txt]: @obastemur emailed you my build log

Comment (obastemur) [inline lib/Runtime/Types/CMakeLists.txt]: @digitalinfinity Thanks! The issue was showing up with debug build. Updated the PR.

**Comment by digitalinfinity** [comment]
Cool- LGTM! Merging

**Pull request merged**